File: sdks/typescript/src/integrations/langchain.ts
/**
 * LangChain.js integration: a callback handler mapping runs to Beater spans.
 *
 *   import { BeaterCallbackHandler } from "@beater/sdk";
 *   await chain.invoke(input, { callbacks: [new BeaterCallbackHandler()] });
 *
 * Implemented structurally (duck-typed) so `@beater/sdk` does not hard-depend on
 * `@langchain/core`. Pass it anywhere a LangChain callback handler is accepted.
 */

import { context, Span, SpanStatusCode, trace } from "@opentelemetry/api";

import { Attr, SpanKind, SpanKindValue } from "../semconv";
import { getTracer } from "../tracing";
import { _internal } from "../observe";

const { toValue } = _internal;

/* eslint-disable @typescript-eslint/no-explicit-any */

export class BeaterCallbackHandler {
  name = "BeaterCallbackHandler";
  awaitHandlers = true;
  private spans = new Map<string, Span>();

  private start(runId: string, parentRunId: string | undefined, name: string, kind: SpanKindValue, input: unknown): void {
    const tracer = getTracer();
    const parent = parentRunId ? this.spans.get(parentRunId) : undefined;
    const ctx = parent ? trace.setSpan(context.active(), parent) : undefined;
    const span = ctx ? tracer.startSpan(name, undefined, ctx) : tracer.startSpan(name);
    span.setAttribute(Attr.SPAN_KIND, kind);
    if (input !== undefined && input !== null) span.setAttribute(Attr.INPUT_VALUE, toValue(input));
    this.spans.set(runId, span);
  }

  private end(runId: string, output?: unknown, error?: unknown): void {
    const span = this.spans.get(runId);
    if (!span) return;
    this.spans.delete(runId);
    if (output !== undefined && output !== null) span.setAttribute(Attr.OUTPUT_VALUE, toValue(output));
    if (error) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: String(error) });
      span.recordException(error as Error);
    } else {
      span.setStatus({ code: SpanStatusCode.OK });
    }
    span.end();
  }

  handleChainStart(chain: any, inputs: any, runId: string, parentRunId?: string): void {
    this.start(runId, parentRunId, chain?.id?.slice?.(-1)?.[0] ?? "chain", SpanKind.AGENT_STEP, inputs);
  }
  handleChainEnd(outputs: any, runId: string): void {
    this.end(runId, outputs);
  }
  handleChainError(err: any, runId: string): void {
    this.end(runId, undefined, err);
  }

  handleLLMStart(llm: any, prompts: any, runId: string, parentRunId?: string): void {
    this.start(runId, parentRunId, "llm", SpanKind.LLM_CALL, prompts);
  }
  handleLLMEnd(output: any, runId: string): void {
    const span = this.spans.get(runId);
    if (span) {
      const usage = output?.llmOutput?.tokenUsage ?? output?.llmOutput?.usage;
      if (usage?.promptTokens != null) span.setAttribute(Attr.LLM_TOKEN_PROMPT, Number(usage.promptTokens));
      if (usage?.completionTokens != null) span.setAttribute(Attr.LLM_TOKEN_COMPLETION, Number(usage.completionTokens));
    }
    let text: string | undefined;
    try {
      text = output?.generations?.[0]?.[0]?.text;
    } catch {
      text = undefined;
    }
    this.end(runId, text);
  }
  handleLLMError(err: any, runId: string): void {
    this.end(runId, undefined, err);
  }

  handleToolStart(tool: any, input: any, runId: string, parentRunId?: string): void {
    this.start(runId, parentRunId, tool?.id?.slice?.(-1)?.[0] ?? "tool", SpanKind.TOOL_CALL, input);
  }
  handleToolEnd(output: any, runId: string): void {
    this.end(runId, output);
  }
  handleToolError(err: any, runId: string): void {
    this.end(runId, undefined, err);
  }

  handleRetrieverStart(retriever: any, query: any, runId: string, parentRunId?: string): void {
    this.start(runId, parentRunId, "retriever", SpanKind.RETRIEVAL_QUERY, query);
  }
  handleRetrieverEnd(documents: any, runId: string): void {
    this.end(runId, documents);
  }
  handleRetrieverError(err: any, runId: string): void {
    this.end(runId, undefined, err);
  }
}

File: sdks/python/beater/integrations/langchain.py
"""LangChain integration: a one-line callback handler.

>>> from beater.integrations.langchain import BeaterCallbackHandler
>>> chain.invoke(inputs, config={"callbacks": [BeaterCallbackHandler()]})
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional
from uuid import UUID

from opentelemetry import context as otel_context
from opentelemetry import trace as otel_trace
from opentelemetry.trace import Status, StatusCode

from ..observe import _to_value
from ..semconv import Attr, SpanKind
from ..tracing import get_tracer

try:  # pragma: no cover - import guard
    from langchain_core.callbacks.base import BaseCallbackHandler
except Exception:  # pragma: no cover - missing or import-incompatible framework
    try:
        from langchain.callbacks.base import BaseCallbackHandler  # type: ignore
    except Exception:  # pragma: no cover
        BaseCallbackHandler = object  # type: ignore


class BeaterCallbackHandler(BaseCallbackHandler):  # type: ignore[misc]
    """Maps LangChain runs (chains, LLMs, tools, retrievers) to Beater spans."""

    def __init__(self) -> None:
        self._spans: Dict[UUID, Any] = {}
        self._tokens: Dict[UUID, Any] = {}

    def _start(self, run_id: UUID, parent_run_id: Optional[UUID], name: str, kind: str, input_value: Any) -> None:
        tracer = get_tracer()
        parent_span = self._spans.get(parent_run_id) if parent_run_id else None
        ctx = otel_trace.set_span_in_context(parent_span) if parent_span else None
        span = tracer.start_span(name, context=ctx)
        span.set_attribute(Attr.SPAN_KIND, kind)
        if input_value is not None:
            span.set_attribute(Attr.INPUT_VALUE, _to_value(input_value))
        self._spans[run_id] = span
        self._tokens[run_id] = otel_context.attach(otel_trace.set_span_in_context(span))

    def _end(self, run_id: UUID, output_value: Any = None, error: Optional[BaseException] = None) -> None:
        span = self._spans.pop(run_id, None)
        token = self._tokens.pop(run_id, None)
        if token is not None:
            otel_context.detach(token)
        if span is None:
            return
        if output_value is not None:
            span.set_attribute(Attr.OUTPUT_VALUE, _to_value(output_value))
        if error is not None:
            span.set_status(Status(StatusCode.ERROR, str(error)))
            span.record_exception(error)
        else:
            span.set_status(Status(StatusCode.OK))
        span.end()

    # -- chains --
    def on_chain_start(self, serialized, inputs, *, run_id, parent_run_id=None, **kwargs):  # type: ignore[no-untyped-def]
        name = (serialized or {}).get("name", "chain") if isinstance(serialized, dict) else "chain"
        self._start(run_id, parent_run_id, name, SpanKind.AGENT_STEP, inputs)

    def on_chain_end(self, outputs, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        self._end(run_id, outputs)

    def on_chain_error(self, error, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        self._end(run_id, error=error)

    # -- LLMs --
    def on_llm_start(self, serialized, prompts, *, run_id, parent_run_id=None, **kwargs):  # type: ignore[no-untyped-def]
        self._start(run_id, parent_run_id, "llm", SpanKind.LLM_CALL, prompts)
        span = self._spans.get(run_id)
        invocation = kwargs.get("invocation_params") or {}
        model = invocation.get("model") or invocation.get("model_name")
        if span is not None and model:
            span.set_attribute(Attr.LLM_MODEL_NAME, str(model))

    def on_chat_model_start(self, serialized, messages, *, run_id, parent_run_id=None, **kwargs):  # type: ignore[no-untyped-def]
        self._start(run_id, parent_run_id, "chat_model", SpanKind.LLM_CALL, messages)

    def on_llm_end(self, response, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        span = self._spans.get(run_id)
        if span is not None:
            usage = {}
            llm_output = getattr(response, "llm_output", None) or {}
            if isinstance(llm_output, dict):
                usage = llm_output.get("token_usage") or llm_output.get("usage") or {}
            if usage.get("prompt_tokens") is not None:
                span.set_attribute(Attr.LLM_TOKEN_PROMPT, int(usage["prompt_tokens"]))
            if usage.get("completion_tokens") is not None:
                span.set_attribute(Attr.LLM_TOKEN_COMPLETION, int(usage["completion_tokens"]))
        output = None
        try:
            output = response.generations[0][0].text
        except (AttributeError, IndexError, TypeError):
            output = None
        self._end(run_id, output)

    def on_llm_error(self, error, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        self._end(run_id, error=error)

    # -- tools --
    def on_tool_start(self, serialized, input_str, *, run_id, parent_run_id=None, **kwargs):  # type: ignore[no-untyped-def]
        name = (serialized or {}).get("name", "tool") if isinstance(serialized, dict) else "tool"
        self._start(run_id, parent_run_id, name, SpanKind.TOOL_CALL, input_str)

    def on_tool_end(self, output, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        self._end(run_id, output)

    def on_tool_error(self, error, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        self._end(run_id, error=error)

    # -- retrievers --
    def on_retriever_start(self, serialized, query, *, run_id, parent_run_id=None, **kwargs):  # type: ignore[no-untyped-def]
        self._start(run_id, parent_run_id, "retriever", SpanKind.RETRIEVAL_QUERY, query)

    def on_retriever_end(self, documents, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        summary = [getattr(doc, "page_content", str(doc))[:200] for doc in (documents or [])]
        self._end(run_id, summary)

    def on_retriever_error(self, error, *, run_id, **kwargs):  # type: ignore[no-untyped-def]
        self._end(run_id, error=error)
